sway-core/src/type_system/info.rs

```diff
@@ -1527,15 +1527,20 @@ impl TypeInfo {
         matches!(self, TypeInfo::UnknownGeneric { .. })
     }
 
-    pub fn new_tuple(engines: &Engines, items: impl IntoIterator<Item = TypeInfo>) -> TypeInfo {
+    pub fn new_tuple(
+        engines: &Engines,
+        items: impl IntoIterator<Item = TypeInfo>,
+        span: Span,
+    ) -> TypeInfo {
+        let source_id = span.source_id();
         let te = engines.te();
         let items = items
             .into_iter()
-            .map(|x| te.insert(engines, x, None))
+            .map(|x| te.insert(engines, x, source_id))
```

Comment:
As commented above, in a general case of usage of `new_tuple`, all different `x` types can be coming from different source files or like `Unknown` and `u64` have no source file attached. Putting the same `source_id` for potentially completely different types out of how I understood the discussion can cause issues in LSP. What I mean is, each individual `source_id` depends on each individual `x` we are inserting, and not on the resulting tuple.

I am still of opinion that the simplest solution in this case, as long as we do not make `insert` more robust in regard to exact semantics of `source_id`, is removing potentially problematic `new_tuple` from `TypeInfo` and putting it to buffer handling, eventually even renaming to `new_buffer_tuple_type` to make it clear it is not generic.

Reply:
Not sure what was the idea then. I am reverting the last commit, and making "new_tuple" specific for the encoding case.